File: backend/app/models/schedule.py
from pydantic import BaseModel
from typing import Optional, Literal, List
from datetime import date


class Dependency(BaseModel):
    predecessor_id: str
    type: Literal["FS", "SS", "FF", "SF"] = "FS"
    lag_days: int = 0


class Activity(BaseModel):
    id: str
    name: str
    wbs_id: str
    duration_days: int
    start_date: Optional[date] = None
    end_date: Optional[date] = None
    predecessors: List[Dependency] = []
    resource: Optional[str] = None
    is_milestone: bool = False
    is_critical: bool = False
    early_start: Optional[date] = None
    early_finish: Optional[date] = None
    late_start: Optional[date] = None
    late_finish: Optional[date] = None
    total_float: Optional[int] = None
    notes: Optional[str] = None


class WBSNode(BaseModel):
    id: str
    name: str
    parent_id: Optional[str] = None
    children: List["WBSNode"] = []
    activities: List[str] = []


WBSNode.model_rebuild()


class Project(BaseModel):
    id: str
    name: str
    description: str
    project_type: str = "residential"
    start_date: date
    wbs: List[WBSNode] = []
    activities: List[Activity] = []
    critical_path: List[str] = []
    project_duration_days: int = 0
    project_end_date: Optional[date] = None


class ProjectSummary(BaseModel):
    id: str
    name: str
    description: str
    activity_count: int
    duration_days: int
    project_type: str

File: frontend/src/types/schedule.ts
export interface Dependency {
  predecessor_id: string;
  type: 'FS' | 'SS' | 'FF' | 'SF';
  lag_days: number;
}

export interface Activity {
  id: string;
  name: string;
  wbs_id: string;
  duration_days: number;
  start_date: string | null;
  end_date: string | null;
  predecessors: Dependency[];
  resource: string | null;
  is_milestone: boolean;
  is_critical: boolean;
  early_start: string | null;
  early_finish: string | null;
  late_start: string | null;
  late_finish: string | null;
  total_float: number | null;
  notes: string | null;
}

export interface WBSNode {
  id: string;
  name: string;
  parent_id: string | null;
  children: WBSNode[];
  activities: string[];
}

export interface Project {
  id: string;
  name: string;
  description: string;
  project_type: string;
  start_date: string;
  wbs: WBSNode[];
  activities: Activity[];
  critical_path: string[];
  project_duration_days: number;
  project_end_date: string | null;
}

export interface ProjectSummary {
  id: string;
  name: string;
  description: string;
  activity_count: number;
  duration_days: number;
  project_type: string;
}
